src/Utilities/Algorithm.hpp

```diff
@@ -0,0 +1,77 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <cstddef>
+#include <numeric>
+
+#include "ErrorHandling/Assert.hpp"
+#include "Utilities/Gsl.hpp"
+
+/*!
+ * \ingroup UtilitiesGroup
+ * \brief Get the indices in the container `t` of the `K` elements that satisfy
+ * `op` after sorting with respect to `op`.
+ *
+ * Specifically, if `op` is `[](const lhs, const rhs) { return lhs < rhs; }`
+ * then the result will be the indices of the `K` smallest elements in `t`. If
+ * `op` is `[](const lhs, const rhs) { return lhs > rhs; }` then the result will
+ * be the indices of the `K` largest elements in `t`.
+ *
+ * \warning The indices are not sorted by `op` when returned. Use `std::sort` to
+ * sort them as follows:
+ * \code
+ * const auto op = [](const double lhs, const double rhs) { return lhs < rhs; };
+ * std::vector<double> t{-1.0, 7.8, 9.3, -22.8, 9.5};
+ * auto result = indices_of<3>(t, op);
+ * std::sort(result.begin(), result.end(),
+ *           [&t, &op](const size_t lhs, const size_t rhs) {
+ *             return op(t[lhs], t[rhs]);
+ *           });
+ * \endcode
+ */
+template <size_t K, typename T, typename Comparator>
+auto indices_of(const T& t, Comparator op) noexcept -> std::array<size_t, K> {
+  ASSERT(K <= t.size(),
+         "The container whose K elements we're finding must have at least size "
+             << K << " but has size " << t.size());
+  std::array<size_t, K> indices{};
+  std::iota(indices.begin(), indices.end(), 0);
+  // Each value in `indices` is an index into `t`. get_index_of_max
+  // returns the index into `indices` corresponding to the largest (per
+  // comparison `op`) value of `t`.
+  const auto get_index_of_max = [&indices, &t, &op ]() noexcept {
+    return std::distance(
+        indices.begin(), std::max_element(indices.begin(), indices.end(), [
+          &t, &op
+        ](const size_t lhs, const size_t rhs) noexcept {
+          return static_cast<size_t>(op(gsl::at(t, lhs), gsl::at(t, rhs)));
```

Comment:
What is happening here? I expect a comparator to return a bool.

Reply:
Perhaps there was a misunderstanding of my previous comment, in which I tried to suggest (but might have been unclear?) changing

```
const auto get_index_of_max = ... { return std::distance ...; }
auto index_of_max = static_cast<size_t>(get_index_of_max());
```
into
```
const auto get_index_of_max = ... { return static_cast<size_t>(std::distance ...); }
auto index_of_max = get_index_of_max();
```
?